src/dependency-manager.cpp

```diff
@@ -0,0 +1,60 @@
+
+#include "dependency-manager.hpp"
+
+void full_dependency_manager_t::node_changed(osmid_t id) {
+    for (auto const way_id : m_object_store->get_ways_by_node(id)) {
+        way_changed(way_id);
+        m_ways_pending_tracker.mark(way_id);
+    }
+
+    for (auto const rel_id : m_object_store->get_rels_by_node(id)) {
+        m_rels_pending_tracker.mark(rel_id);
+    }
+}
+
+void full_dependency_manager_t::way_changed(osmid_t id) {
+    if (m_ways_pending_tracker.is_marked(id)) {
+        return;
+    }
+
+    for (auto const rel_id : m_object_store->get_rels_by_way(id)) {
+        m_rels_pending_tracker.mark(rel_id);
+    }
+}
+
+void full_dependency_manager_t::relation_changed(osmid_t id) {
+    if (m_rels_pending_tracker.is_marked(id)) {
+        return;
```

Comment:
This is not true. A marked relation has not necessarily marked its dependencies. That would only be true if we called relation_changed in lines 11, 21 and 31.

Reply:
fixed